casper/src/main/scala/coop/rchain/casper/api/MachineVerifiableDag.scala

```diff
@@ -0,0 +1,44 @@
+package coop.rchain.casper.api
+
+import coop.rchain.casper._
+import coop.rchain.casper.protocol._
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.blockstorage.BlockStore
+import cats._, cats.data._, cats.implicits._
+
+final case class VerifiableEdge(from: String, to: String)
+
+object VerifiableEdge {
+
+  implicit def showBlockHash: Show[BlockHash] = new Show[BlockHash] {
+    def show(blockHash: BlockHash): String =
+      Base16.encode(blockHash.toByteArray)
+  }
+
+  implicit def showVerifiableEdge: Show[VerifiableEdge] = new Show[VerifiableEdge] {
+    def show(ve: VerifiableEdge): String = s"${ve.from} ${ve.to}"
+  }
+}
+
+object MachineVerifiableDag {
+  def apply[F[_]: Monad](
+      toposort: TopoSort,
+      fetchParents: BlockHash => F[List[BlockHash]]
+  ): F[List[VerifiableEdge]] = {
+    import VerifiableEdge._
+
+    toposort
+      .foldM(List.empty[VerifiableEdge]) {
+        case (acc, blockHashes) =>
+          for {
+            parents          <- blockHashes.toList.traverse(fetchParents)
+            blocks           = blockHashes.toList.map(_.show)
+            blocksAndParents = blocks.zip(parents.map(_.map(p => p.show)))
+            entries = blocksAndParents.flatMap {
```

Comment:
I think this can be simplified. The zip seems unnecessary. Something like the code below should also work but I haven't tried to compile it

```scala
blockHashes.toStream
    .traverse ( h=> fetchParents(h).map(p => (h.show, p.show)))
    .map(blocksAndParents => 
        val entries = blockAndParents.flatMap {...}
        entries ++ acc
    )
...
```

Reply:
👍 not sure, if more readable but commit was added